File: src/dotnet/PolyBus/Transport/InMemory/InMemoryMessageBroker.cs
using System.Collections.Concurrent;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using PolyBus.Transport.Transactions;
using PolyBus.Transport.Transactions.Messages;

namespace PolyBus.Transport.InMemory;

/// <summary>
/// A message broker that uses in-memory transport for message passing.
/// </summary>
public class InMemoryMessageBroker
{
    /// <summary>
    /// The collection of in-memory endpoints managed by this broker.
    /// </summary>
    public ConcurrentDictionary<string, InMemoryEndpoint> Endpoints { get; } = new();

    /// <summary>
    /// The logger for the InMemoryMessageBroker.
    /// </summary>
    public ILogger<InMemoryMessageBroker> Log { get; set; } = NullLogger<InMemoryMessageBroker>.Instance;

    /// <summary>
    /// The ITransport factory method.
    /// </summary>
    public Task<ITransport> AddEndpoint(PolyBusBuilder builder, IPolyBus bus)
    {
        var endpoint = new InMemoryEndpoint(this, bus);
        Endpoints[bus.Name] = endpoint;
        return Task.FromResult<ITransport>(endpoint);
    }

    /// <summary>
    /// Processes the transaction and distributes outgoing messages to the appropriate endpoints.
    /// </summary>
    public async void Send(Transaction transaction)
    {
        if (transaction.OutgoingMessages.Count == 0)
        {
            return;
        }

        try
        {
            await Task.Yield();

            if (Interlocked.Increment(ref _count) == 1)
            {
                while (_emptySignal.CurrentCount > 0)
                {
                    await _emptySignal.WaitAsync(0);
                }
            }

            var tasks = new List<Task>();
            var now = DateTime.UtcNow;

            foreach (var message in transaction.OutgoingMessages)
            {
                foreach (var endpoint in Endpoints.Values)
                {
                    var isDeadLetter = endpoint.DeadLetterEndpoint == message.Endpoint;
                    if (isDeadLetter
                        || endpoint.Bus.Name == message.Endpoint
                        || (message.Endpoint == null
                            && (message.MessageInfo.Endpoint == endpoint.Bus.Name
                                || endpoint.IsSubscribed(message.MessageInfo))))
                    {
                        var incomingMessage = new IncomingMessage(endpoint.Bus, message.Body, message.MessageInfo)
                        {
                            Headers = new Dictionary<string, string>(message.Headers)
                        };
                        if (message.DeliverAt != null)
                        {
                            var wait = message.DeliverAt.Value - now;
                            if (wait > TimeSpan.Zero)
                            {
                                DelayedSend(endpoint, incomingMessage, wait, isDeadLetter);
                                continue;
                            }
                        }

                        var task = endpoint.HandleMessage(incomingMessage, isDeadLetter);
                        tasks.Add(task);
                    }
                }
            }

            await Task.WhenAll(tasks);
        }
        catch (Exception error)
        {
            Log.LogError(error, error.Message);
        }
        finally
        {
            if (Interlocked.Decrement(ref _count) == 0)
            {
                _emptySignal.Release();
            }
        }
    }
    async void DelayedSend(InMemoryEndpoint endpoint, IncomingMessage message, TimeSpan delay, bool isDeadLetter)
    {
        try
        {
            await Task.Delay(delay, _cts.Token);
            await endpoint.HandleMessage(message, isDeadLetter);
        }
        catch (OperationCanceledException)
        {
            // Ignore cancellation
        }
        catch (Exception error)
        {
            Log.LogError(error, error.Message);
        }
    }

    /// <summary>
    /// Stops all endpoints and waits for in-flight messages to be processed.
    /// </summary>
    public async Task Stop()
    {
        foreach (var endpoint in Endpoints.Values)
        {
            await endpoint.Stop();
        }
        _cts.Cancel();
        if (Volatile.Read(ref _count) > 0)
        {
            await _emptySignal.WaitAsync();
        }
    }
    int _count;
    readonly CancellationTokenSource _cts = new();
    readonly SemaphoreSlim _emptySignal = new(0, 1);
}

File: src/kotlin/src/main/kotlin/polybus/transport/inmemory/InMemoryMessageBroker.kt
package polybus.transport.inmemory

import polybus.IPolyBus
import polybus.PolyBusBuilder
import polybus.transport.ITransport
import polybus.transport.transactions.Transaction
import polybus.transport.transactions.messages.IncomingMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Level
import java.util.logging.Logger

open class InMemoryMessageBroker {
    val endpoints: ConcurrentHashMap<String, InMemoryEndpoint> = ConcurrentHashMap()
    var log: Logger = Logger.getLogger(InMemoryMessageBroker::class.java.name)

    private val _scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val _activeSends = AtomicInteger(0)
    private val _jobs: ConcurrentHashMap.KeySetView<kotlinx.coroutines.Job, Boolean> = ConcurrentHashMap.newKeySet()

    open suspend fun addEndpoint(builder: PolyBusBuilder, bus: IPolyBus): ITransport {
        val endpoint = InMemoryEndpoint(this, bus)
        endpoints[bus.name] = endpoint
        return endpoint
    }

    open fun send(transaction: Transaction) {
        if (transaction.outgoingMessages.isEmpty()) {
            return
        }

        val job = _scope.launch {
            _activeSends.incrementAndGet()
            try {
                val now = Instant.now()
                transaction.outgoingMessages.forEach { message ->
                    endpoints.values.forEach { endpoint ->
                        val isDeadLetter = endpoint.deadLetterEndpoint == message.endpoint
                        val shouldSend = isDeadLetter ||
                            endpoint.bus.name == message.endpoint ||
                            (message.endpoint == null &&
                                (message.messageInfo.endpoint == endpoint.bus.name ||
                                    endpoint.isSubscribed(message.messageInfo)))

                        if (!shouldSend) {
                            return@forEach
                        }

                        val incomingMessage = IncomingMessage(endpoint.bus, message.body, message.messageInfo)
                        incomingMessage.headers = message.headers.toMutableMap()

                        val deliverAt = message.deliverAt
                        if (deliverAt != null) {
                            val wait = Duration.between(now, deliverAt)
                            if (!wait.isNegative && !wait.isZero) {
                                delayedSend(endpoint, incomingMessage, wait, isDeadLetter)
                                return@forEach
                            }
                        }

                        endpoint.handleMessage(incomingMessage, isDeadLetter)
                    }
                }
            } catch (error: Throwable) {
                log.log(Level.SEVERE, error.message, error)
            } finally {
                _activeSends.decrementAndGet()
            }
        }

        _jobs.add(job)
        job.invokeOnCompletion { _jobs.remove(job) }
    }

    private fun delayedSend(
        endpoint: InMemoryEndpoint,
        message: IncomingMessage,
        delayBy: Duration,
        isDeadLetter: Boolean
    ) {
        val job = _scope.launch {
            try {
                delay(delayBy.toMillis())
                endpoint.handleMessage(message, isDeadLetter)
            } catch (_: CancellationException) {
            } catch (error: Throwable) {
                log.log(Level.SEVERE, error.message, error)
            }
        }

        _jobs.add(job)
        job.invokeOnCompletion { _jobs.remove(job) }
    }

    open suspend fun stop() {
        endpoints.values.forEach { it.stop() }
        _jobs.forEach { it.cancel() }
        _jobs.toList().forEach { it.join() }
        while (_activeSends.get() > 0) {
            delay(10)
        }
    }
}
